File: templates/skills/workflow/scripts/quality_review.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
quality_review 执行侧 helper。

沉淀 execution side 两阶段审查的稳定数据结构与状态写入逻辑，
供 execute 阶段与后续 execution_sequencer 复用。
"""

from __future__ import annotations

import json
import sys
from datetime import datetime
from typing import Any, Dict, List, Optional

from state_manager import read_state, write_state
from verification import create_evidence
from workflow_types import ensure_state_defaults, get_review_result

GATE_BUDGET = {
    "max_total_loops": 4,
    "max_diff_context_chars": 50000,
    "cache_stage1": True,
}


def iso_now() -> str:
    return datetime.now().isoformat()


def create_review_subject(
    base_commit: str,
    requirement_ids: Optional[List[str]] = None,
    critical_constraints: Optional[List[str]] = None,
) -> Dict[str, Any]:
    return {
        "kind": "diff_window",
        "ref": f"{base_commit}..HEAD",
        "requirement_ids": requirement_ids or [],
        "critical_constraints": critical_constraints or [],
    }


def create_diff_window(
    base_commit: str,
    from_task: Optional[str] = None,
    to_task: Optional[str] = None,
    files_changed: int = 0,
) -> Dict[str, Any]:
    return {
        "base_commit": base_commit,
        "from_task": from_task,
        "to_task": to_task,
        "files_changed": files_changed,
    }


def extract_issue_count(result: Optional[Dict[str, Any]]) -> int:
    if not result:
        return 0

    total = 0
    for key in ("missing", "extra", "misunderstandings", "coverage_gaps", "blocking_issues"):
        value = result.get(key)
        if isinstance(value, list):
            total += len(value)

    issues = result.get("issues")
    if isinstance(issues, dict):
        for key in ("critical", "important", "minor"):
            value = issues.get(key)
            if isinstance(value, list):
                total += len(value)

    return total


def collect_blocking_issues(result: Optional[Dict[str, Any]]) -> List[Dict[str, Any]]:
    if not result:
        return []

    blocking = result.get("blocking_issues")
    if isinstance(blocking, list):
        return blocking

    collected: List[Dict[str, Any]] = []
    for key in ("missing", "extra", "misunderstandings", "coverage_gaps"):
        value = result.get(key)
        if not isinstance(value, list):
            continue
        for item in value:
            if isinstance(item, dict):
                collected.append(item)
            else:
                collected.append({"description": str(item)})

    issues = result.get("issues")
    if isinstance(issues, dict):
        for level in ("critical", "important"):
            value = issues.get(level)
            if not isinstance(value, list):
                continue
            for item in value:
                if isinstance(item, dict):
                    collected.append(item)
                else:
                    collected.append({"description": str(item), "severity": level})

    return collected


def build_pass_gate_result(
    task_id: str,
    base_commit: str,
    current_commit: Optional[str] = None,
    from_task: Optional[str] = None,
    to_task: Optional[str] = None,
    files_changed: int = 0,
    requirement_ids: Optional[List[str]] = None,
    critical_constraints: Optional[List[str]] = None,
    stage1_attempts: int = 1,
    stage2_attempts: int = 1,
    stage1_issues_found: int = 0,
    critical_count: int = 0,
    important_count: int = 0,
    minor_count: int = 0,
    reviewer: str = "subagent",
) -> Dict[str, Any]:
    attempts = stage1_attempts + stage2_attempts
    now = iso_now()
    return {
        "review_type": "quality_review",
        "review_mode": "machine_loop",
        "gate_task_id": task_id,
        "subject": create_review_subject(base_commit, requirement_ids, critical_constraints),
        "max_attempts": GATE_BUDGET["max_total_loops"],
        "attempt": attempts,
        "last_decision": "pass",
        "next_action": "continue_execution",
        "commit_hash": current_commit or base_commit,
        "diff_window": create_diff_window(base_commit, from_task, to_task, files_changed),
        "stage1": {
            "passed": True,
            "attempts": stage1_attempts,
            "issues_found": stage1_issues_found,
            "completed_at": now,
        },
        "stage2": {
            "passed": True,
            "attempts": stage2_attempts,
            "assessment": "approved",
            "critical_count": critical_count,
            "important_count": important_count,
            "minor_count": minor_count,
            "completed_at": now,
        },
        "overall_passed": True,
        "reviewed_at": now,
        "reviewer": reviewer,
    }


def build_failed_gate_result(
    task_id: str,
    failed_stage: str,
    base_commit: str,
    current_commit: Optional[str] = None,
    from_task: Optional[str] = None,
    to_task: Optional[str] = None,
    files_changed: int = 0,
    requirement_ids: Optional[List[str]] = None,
    critical_constraints: Optional[List[str]] = None,
    stage1_attempts: int = 1,
    total_attempts: int = 1,
    last_result: Optional[Dict[str, Any]] = None,
    reviewer: str = "subagent",
) -> Dict[str, Any]:
    budget_exhausted = total_attempts > GATE_BUDGET["max_total_loops"]
    terminal_decision = "rejected" if budget_exhausted or failed_stage == "stage2" else "revise"
    next_action = "mark_task_failed_or_escalate" if terminal_decision == "rejected" else "fix_and_retry_or_escalate"
    now = iso_now()

    result: Dict[str, Any] = {
        "review_type": "quality_review",
        "review_mode": "machine_loop",
        "gate_task_id": task_id,
        "subject": create_review_subject(base_commit, requirement_ids, critical_constraints),
        "max_attempts": GATE_BUDGET["max_total_loops"],
        "attempt": total_attempts,
        "last_decision": terminal_decision,
        "next_action": next_action,
        "blocking_issues": collect_blocking_issues(last_result),
        "reviewed_at": now,
        "reviewer": reviewer,
        "commit_hash": current_commit or base_commit,
        "diff_window": create_diff_window(base_commit, from_task, to_task, files_changed),
        "stage1": {
            "passed": failed_stage != "stage1",
            "attempts": stage1_attempts,
            "issues_found": extract_issue_count(last_result),
            "completed_at": now,
        },
        "overall_passed": False,
    }

    if failed_stage != "stage1":
        issues = last_result.get("issues") if isinstance(last_result, dict) else {}
        result["stage2"] = {
            "passed": False,
            "attempts": max(total_attempts - stage1_attempts, 0),
            "assessment": (last_result or {}).get("assessment", "rejected"),
            "critical_count": len((issues or {}).get("critical", []) or []),
            "important_count": len((issues or {}).get("important", []) or []),
            "minor_count": len((issues or {}).get("minor", []) or []),
            "completed_at": now,
        }

    return result


def write_quality_gate_result(
    state_path: str,
    task_id: str,
    gate_result: Dict[str, Any],
) -> Dict[str, Any]:
    state = ensure_state_defaults(read_state(state_path))
    quality_gates = state.setdefault("quality_gates", {})
    quality_gates[task_id] = gate_result
    write_state(state_path, state)
    return gate_result


def read_quality_gate_result(state_path: str, task_id: str) -> Optional[Dict[str, Any]]:
    state = ensure_state_defaults(read_state(state_path))
    return get_review_result(state, task_id)


def create_quality_review_evidence(task_id: str, gate_result: Dict[str, Any]) -> Dict[str, Any]:
    passed = bool(gate_result.get("overall_passed"))
    stage1 = gate_result.get("stage1") or {}
    stage2 = gate_result.get("stage2") or {}
    output_summary = (
        f"Stage 1 passed={stage1.get('passed', False)} attempts={stage1.get('attempts', 0)}, "
        f"Stage 2 passed={stage2.get('passed', False)} attempts={stage2.get('attempts', 0)}, "
        f"decision={gate_result.get('last_decision')}"
    )
    return create_evidence(
        command="two-stage code review",
        exit_code=0 if passed else 1,
        output_summary=output_summary,
        passed=passed,
        artifact_ref=f"quality_gates.{task_id}",
    )


def main() -> int:
    import argparse

    parser = argparse.ArgumentParser(description="quality_review 执行侧 helper")
    sub = parser.add_subparsers(dest="command")

    p_pass = sub.add_parser("pass", help="生成通过态 quality gate 结果")
    p_pass.add_argument("task_id")
    p_pass.add_argument("--base-commit", required=True)
    p_pass.add_argument("--current-commit")
    p_pass.add_argument("--from-task")
    p_pass.add_argument("--to-task")
    p_pass.add_argument("--files-changed", type=int, default=0)
    p_pass.add_argument("--stage1-attempts", type=int, default=1)
    p_pass.add_argument("--stage2-attempts", type=int, default=1)
    p_pass.add_argument("--stage1-issues-found", type=int, default=0)
    p_pass.add_argument("--critical-count", type=int, default=0)
    p_pass.add_argument("--important-count", type=int, default=0)
    p_pass.add_argument("--minor-count", type=int, default=0)
    p_pass.add_argument("--requirement-ids", default="")
    p_pass.add_argument("--critical-constraints", default="")
    p_pass.add_argument("--reviewer", default="subagent")
    p_pass.add_argument("--state-file")

    p_fail = sub.add_parser("fail", help="生成失败态 quality gate 结果")
    p_fail.add_argument("task_id")
    p_fail.add_argument("--failed-stage", choices=["stage1", "stage2", "stage1_recheck"], required=True)
    p_fail.add_argument("--base-commit", required=True)
    p_fail.add_argument("--current-commit")
    p_fail.add_argument("--from-task")
    p_fail.add_argument("--to-task")
    p_fail.add_argument("--files-changed", type=int, default=0)
    p_fail.add_argument("--stage1-attempts", type=int, default=1)
    p_fail.add_argument("--total-attempts", type=int, default=1)
    p_fail.add_argument("--requirement-ids", default="")
    p_fail.add_argument("--critical-constraints", default="")
    p_fail.add_argument("--reviewer", default="subagent")
    p_fail.add_argument("--last-result-json", default="{}")
    p_fail.add_argument("--state-file")

    p_read = sub.add_parser("read", help="读取 quality gate 结果")
    p_read.add_argument("state_file")
    p_read.add_argument("task_id")

    sub.add_parser("budget", help="读取 quality review 预算")

    args = parser.parse_args()
    split = lambda s: [item.strip() for item in s.split(",") if item.strip()]

    if args.command == "pass":
        gate_result = build_pass_gate_result(
            args.task_id,
            args.base_commit,
            current_commit=args.current_commit,
            from_task=args.from_task,
            to_task=args.to_task,
            files_changed=args.files_changed,
            requirement_ids=split(args.requirement_ids),
            critical_constraints=split(args.critical_constraints),
            stage1_attempts=args.stage1_attempts,
            stage2_attempts=args.stage2_attempts,
            stage1_issues_found=args.stage1_issues_found,
            critical_count=args.critical_count,
            important_count=args.important_count,
            minor_count=args.minor_count,
            reviewer=args.reviewer,
        )
        if args.state_file:
            write_quality_gate_result(args.state_file, args.task_id, gate_result)
        print(json.dumps({"gate_result": gate_result, "evidence": create_quality_review_evidence(args.task_id, gate_result)}, ensure_ascii=False))
        return 0

    if args.command == "fail":
        try:
            last_result = json.loads(args.last_result_json)
        except json.JSONDecodeError as error:
            print(json.dumps({"error": f"invalid last-result-json: {error}"}, ensure_ascii=False))
            return 1
        gate_result = build_failed_gate_result(
            args.task_id,
            args.failed_stage,
            args.base_commit,
            current_commit=args.current_commit,
            from_task=args.from_task,
            to_task=args.to_task,
            files_changed=args.files_changed,
            requirement_ids=split(args.requirement_ids),
            critical_constraints=split(args.critical_constraints),
            stage1_attempts=args.stage1_attempts,
            total_attempts=args.total_attempts,
            last_result=last_result,
            reviewer=args.reviewer,
        )
        if args.state_file:
            write_quality_gate_result(args.state_file, args.task_id, gate_result)
        print(json.dumps({"gate_result": gate_result, "evidence": create_quality_review_evidence(args.task_id, gate_result)}, ensure_ascii=False))
        return 0

    if args.command == "read":
        print(json.dumps({"review": read_quality_gate_result(args.state_file, args.task_id)}, ensure_ascii=False))
        return 0

    if args.command == "budget":
        print(json.dumps(GATE_BUDGET, ensure_ascii=False))
        return 0

    parser.print_help()
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: core/utils/workflow/quality_review.js
#!/usr/bin/env node

const { spawnSync } = require('child_process')
const { resolveStatePath } = require('./state_manager')
const { assertCanonicalWorkflowStatePath } = require('./path_utils')
const { ensureStateDefaults } = require('./workflow_types')
const { buildInjectedContext, buildAgentPrompt, classifyRoleSignals, resolveRoleProfile, STAGE2_REVIEW_MODE_SET } = require('./role_injection')

// per-task gate 持久化 + governor budget 已退役（R-002）。
// 末尾终审复用的 buildPass/FailGateResult 仍需 review loop 上限来整形结果，
// 保留为不导出的内部常量，不再承担 budget 决策。
// 4 = 1 次首审 + 3 轮重派，对应 SKILL Step 4.2 / subagent-driven.md 的「循环上限 3 轮」（attempts 计数，非轮数）。
const MAX_REVIEW_LOOPS = 4

function normalizeStage2ReviewMode(value, fallback = 'single_reviewer') {
  if (!value) return fallback
  const str = String(value).trim()
  if (STAGE2_REVIEW_MODE_SET.has(str)) return str
  process.stderr.write(`[quality_review] unknown --review-mode "${str}", falling back to "${fallback}"\n`)
  return fallback
}

function isoNow() {
  return new Date().toISOString()
}

function createReviewSubject(baseCommit, requirementIds = [], criticalConstraints = []) {
  const ref = baseCommit ? `${baseCommit}..HEAD` : 'HEAD'
  return {
    kind: 'diff_window',
    ref,
    requirement_ids: requirementIds,
    critical_constraints: criticalConstraints,
  }
}

function createDiffWindow(baseCommit, fromTask = null, toTask = null, filesChanged = 0) {
  return {
    base_commit: baseCommit,
    from_task: fromTask,
    to_task: toTask,
    files_changed: filesChanged,
  }
}

function getGitHead(projectRoot = process.cwd()) {
  const result = spawnSync('git', ['rev-parse', 'HEAD'], {
    cwd: projectRoot,
    encoding: 'utf8',
    timeout: 5000,
  })
  if (result.status !== 0) return null
  const commit = String(result.stdout || '').trim()
  return commit || null
}

function resolveQualityReviewProfile(state = {}, requirementIds = [], criticalConstraints = [], diffWindow = null) {
  const normalized = ensureStateDefaults(state)
  const existing = (((normalized.context_injection || {}).execution) || {}).quality_review_stage2 || {}
  const signals = classifyRoleSignals('', [], null, {
    requirementIds,
    criticalConstraints,
    summary: requirementIds.join(' '),
    taskName: criticalConstraints.join(' '),
  })
  if (existing.profile || existing.role) {
    const profile = {
      ...resolveRoleProfile('quality_review_stage2', signals, normalized.collaboration || {}, normalized.sessions || {}),
      role: existing.role || resolveRoleProfile('quality_review_stage2', signals, normalized.collaboration || {}, normalized.sessions || {}).role,
      profile: existing.profile || resolveRoleProfile('quality_review_stage2', signals, normalized.collaboration || {}, normalized.sessions || {}).profile,
    }
    return {
      signals,
      profile,
      injectedContext: buildInjectedContext(
        createReviewSubject(diffWindow?.base_commit || 'HEAD', requirementIds, criticalConstraints),
        profile,
        signals,
        { diff_window: diffWindow }
      ),
    }
  }
  const profile = resolveRoleProfile('quality_review_stage2', signals, normalized.collaboration || {}, normalized.sessions || {})
  return {
    signals,
    profile,
    injectedContext: buildInjectedContext(
      createReviewSubject(diffWindow?.base_commit || 'HEAD', requirementIds, criticalConstraints),
      profile,
      signals,
      { diff_window: diffWindow }
    ),
  }
}

function createReviewerPrompt(state = {}, requirementIds = [], criticalConstraints = [], diffWindow = null) {
  const { profile, injectedContext } = resolveQualityReviewProfile(state, requirementIds, criticalConstraints, diffWindow)
  return {
    role: profile.role,
    profile: profile.profile,
    prompt: buildAgentPrompt(profile, injectedContext, ((state || {}).sessions || {}).platform || 'claude-code'),
    injected_context: injectedContext,
  }
}

function extractIssueCount(result) {
  if (!result) return 0
  let total = 0
  for (const key of ['missing', 'extra', 'misunderstandings', 'coverage_gaps', 'blocking_issues']) {
    if (Array.isArray(result[key])) total += result[key].length
  }
  if (result.issues && typeof result.issues === 'object') {
    for (const key of ['critical', 'important', 'minor']) {
      if (Array.isArray(result.issues[key])) total += result.issues[key].length
    }
  }
  // Probe E 只在 cross_layer_depth_gap 上携带细节。如果 blocking_issues 已经包含了该条目，上面的计数已经算过；
  // 否则（例如只传 crossLayerDepthGap 走 options 路径）按 missing_sections 的条数计入，退化最少为 1。
  if (result.cross_layer_depth_gap && result.cross_layer_depth_gap.triggered) {
    const hasInBlocking = Array.isArray(result.blocking_issues)
      && result.blocking_issues.some((x) => x && x.type === 'cross_layer_depth_gap')
    if (!hasInBlocking) {
      const sections = result.cross_layer_depth_gap.missing_sections
      total += Array.isArray(sections) && sections.length > 0 ? sections.length : 1
    }
  }
  return total
}

function buildCrossLayerDepthGapIssue(gap) {
  return {
    type: 'cross_layer_depth_gap',
    severity: 'critical',
    description: gap.description || 'Infra/cross-layer 改动命中 Probe E，但关联 code-spec 缺少关键段（Validation & Error Matrix / Good-Base-Bad Cases / Tests Required）。',
    files: Array.isArray(gap.files) ? [...gap.files] : [],
    specs: Array.isArray(gap.specs) ? [...gap.specs] : [],
    missing_sections: Array.isArray(gap.missing_sections) ? [...gap.missing_sections] : [],
  }
}

function collectBlockingIssues(result) {
  if (!result) return []
  // Probe E 条目必须**追加**而不是重建：外部 reviewer 产出的 `blocking_issues` 数组要原样保留，Probe E 只在命中时补一条。
  // 因此即使 result.blocking_issues 已是数组，也要先拷一份再追加，不能早退把 Probe E 吞掉。
  if (Array.isArray(result.blocking_issues)) {
    const base = [...result.blocking_issues]
    if (result.cross_layer_depth_gap && typeof result.cross_layer_depth_gap === 'object'
        && !base.some((x) => x && x.type === 'cross_layer_depth_gap')) {
      base.push(buildCrossLayerDepthGapIssue(result.cross_layer_depth_gap))
    }
    return base
  }
  const collected = []
  for (const key of ['missing', 'extra', 'misunderstandings', 'coverage_gaps']) {
    if (!Array.isArray(result[key])) continue
    for (const item of result[key]) collected.push(typeof item === 'object' ? item : { description: String(item) })
  }
  if (result.issues && typeof result.issues === 'object') {
    for (const level of ['critical', 'important']) {
      if (!Array.isArray(result.issues[level])) continue
      for (const item of result.issues[level]) collected.push(typeof item === 'object' ? item : { description: String(item), severity: level })
    }
  }
  if (result.cross_layer_depth_gap && typeof result.cross_layer_depth_gap === 'object') {
    collected.push(buildCrossLayerDepthGapIssue(result.cross_layer_depth_gap))
  }
  return collected
}

// 归一化 Stage 1 的 code_specs_check 子段。per-change 诊断永远 advisory，只记录计数与是否执行过。
function normalizeCodeSpecsCheck({ performed = true, findingsCount = 0 } = {}) {
  return {
    performed: Boolean(performed),
    advisory: true,
    findings_count: Number.isFinite(findingsCount) ? Math.max(0, Math.trunc(findingsCount)) : 0,
  }
}

// 归一化 Probe E 的阻塞标记。只在命中时写入；未命中则 stage1 子段中不出现该字段。
function normalizeCrossLayerDepthGap({ triggered = false, files = [], specs = [], missingSections = [], description = '' } = {}) {
  return {
    triggered: Boolean(triggered),
    files: Array.isArray(files) ? [...files] : [],
    specs: Array.isArray(specs) ? [...specs] : [],
    missing_sections: Array.isArray(missingSections) ? [...missingSections] : [],
    description: description ? String(description) : '',
  }
}

function buildPassGateResult(taskId, baseCommit, currentCommit = null, fromTask = null, toTask = null, filesChanged = 0, requirementIds = [], criticalConstraints = [], stage1Attempts = 1, stage2Attempts = 1, stage1IssuesFound = 0, criticalCount = 0, importantCount = 0, minorCount = 0, reviewer = 'subagent', state = {}, stage2ReviewMode = 'single_reviewer', stage2CodexStatus = null, stage2ReviewCycleId = null, options = {}) {
  const attempts = stage1Attempts + stage2Attempts
  const now = isoNow()
  const diffWindow = createDiffWindow(baseCommit, fromTask, toTask, filesChanged)
  const reviewerPrompt = createReviewerPrompt(state, requirementIds, criticalConstraints, diffWindow)
  const stage1 = { passed: true, attempts: stage1Attempts, issues_found: stage1IssuesFound, completed_at: now }
  // Code Specs Check 仅 advisory，只要 CLI 有声明就写入；pass 路径不应命中 cross_layer_depth_gap，保留字段以便下游统一读取。
  stage1.code_specs_check = normalizeCodeSpecsCheck(options.codeSpecsCheck || {})
  return {
    review_type: 'quality_review',
    review_mode: 'machine_loop',
    gate_task_id: taskId,
    subject: createReviewSubject(baseCommit, requirementIds, criticalConstraints),
    max_attempts: MAX_REVIEW_LOOPS,
    attempt: attempts,
    last_decision: 'pass',
    next_action: 'continue_execution',
    commit_hash: currentCommit || baseCommit,
    diff_window: diffWindow,
    stage1,
    stage2: { passed: true, attempts: stage2Attempts, assessment: 'approved', critical_count: criticalCount, important_count: importantCount, minor_count: minorCount, completed_at: now, role: reviewerPrompt.role, profile: reviewerPrompt.profile, review_mode: stage2ReviewMode, codex_status: stage2CodexStatus, review_cycle_id: stage2ReviewCycleId, raw_results: null, merged: null },
    overall_passed: true,
    reviewed_at: now,
    reviewer,
  }
}

function buildFailedGateResult(taskId, failedStage, baseCommit, currentCommit = null, fromTask = null, toTask = null, filesChanged = 0, requirementIds = [], criticalConstraints = [], stage1Attempts = 1, totalAttempts = 1, lastResult = null, reviewer = 'subagent', state = {}, options = {}) {
  const budgetExhausted = totalAttempts >= MAX_REVIEW_LOOPS
  const stage1Failed = failedStage === 'stage1' || failedStage === 'stage1_recheck'
  const stage2Failed = failedStage === 'stage2'
  const terminalDecision = budgetExhausted || stage2Failed ? 'rejected' : 'revise'
  const nextAction = terminalDecision === 'rejected' ? 'mark_task_failed_or_escalate' : 'fix_and_retry_or_escalate'
  const now = isoNow()
  const diffWindow = createDiffWindow(baseCommit, fromTask, toTask, filesChanged)
  const reviewerPrompt = createReviewerPrompt(state, requirementIds, criticalConstraints, diffWindow)
  // 把上游 CLI 传入的 cross_layer_depth_gap / code_specs_check 合并进 lastResult，再走 collectBlockingIssues。
  const mergedLastResult = { ...(lastResult && typeof lastResult === 'object' ? lastResult : {}) }
  if (options.crossLayerDepthGap && options.crossLayerDepthGap.triggered) {
    mergedLastResult.cross_layer_depth_gap = normalizeCrossLayerDepthGap(options.crossLayerDepthGap)
  }
  const stage1 = { passed: !stage1Failed, attempts: stage1Attempts, issues_found: extractIssueCount(mergedLastResult), completed_at: now }
  stage1.code_specs_check = normalizeCodeSpecsCheck(options.codeSpecsCheck || {})
  if (mergedLastResult.cross_layer_depth_gap) stage1.cross_layer_depth_gap = mergedLastResult.cross_layer_depth_gap
  const result = {
    review_type: 'quality_review',
    review_mode: 'machine_loop',
    gate_task_id: taskId,
    subject: createReviewSubject(baseCommit, requirementIds, criticalConstraints),
    max_attempts: MAX_REVIEW_LOOPS,
    attempt: totalAttempts,
    last_decision: terminalDecision,
    next_action: nextAction,
    blocking_issues: collectBlockingIssues(mergedLastResult),
    reviewed_at: now,
    reviewer,
    commit_hash: currentCommit || baseCommit,
    diff_window: diffWindow,
    stage1,
    overall_passed: false,
  }
  if (stage2Failed) {
    const issues = typeof lastResult === 'object' && lastResult ? (lastResult.issues || {}) : {}
    result.stage2 = {
      passed: false,
      attempts: Math.max(totalAttempts - stage1Attempts, 0),
      assessment: (lastResult || {}).assessment || 'rejected',
      critical_count: (issues.critical || []).length,
      important_count: (issues.important || []).length,
      minor_count: (issues.minor || []).length,
      completed_at: now,
      role: reviewerPrompt.role,
      profile: reviewerPrompt.profile,
      review_mode: normalizeStage2ReviewMode((lastResult || {}).review_mode),
      codex_status: (lastResult || {}).codex_status || null,
      review_cycle_id: (lastResult || {}).review_cycle_id || null,
      raw_results: (lastResult || {}).raw_results || null,
      merged: (lastResult || {}).merged || null,
    }
  } else if (failedStage === 'stage1_recheck') {
    result.stage2 = {
      passed: true,
      attempts: Math.max(totalAttempts - stage1Attempts, 0),
      assessment: 'approved',
      critical_count: 0,
      important_count: 0,
      minor_count: 0,
      completed_at: now,
      role: reviewerPrompt.role,
      profile: reviewerPrompt.profile,
      review_mode: 'single_reviewer',
      codex_status: null,
      review_cycle_id: null,
      raw_results: null,
      merged: null,
    }
  }
  return result
}

function resolveCliStatePath(projectId = null, stateFile = null) {
  if (stateFile) return assertCanonicalWorkflowStatePath(stateFile, projectId)
  if (projectId) return resolveStatePath(projectId)
  throw new Error('missing state reference')
}

function resolveExistingCliStatePath(projectId = null, stateFile = null) {
  try {
    const statePath = resolveCliStatePath(projectId, stateFile)
    return require('fs').existsSync(statePath) ? statePath : null
  } catch {
    return null
  }
}

// R-002：per-task gate 持久化（pass/fail 落盘、read 读盘）与 governor budget 子命令均已退役。
// 本模块现在仅作为库导出 reviewer prompt 构造 + review 结果整形辅助，供末尾终审复用；无 CLI 子命令。
function main() {
  process.stderr.write('quality_review.js 无 CLI 子命令（per-task gate 持久化已退役，模块仅供库内引用）\n')
  process.exitCode = 1
}

module.exports = {
  isoNow,
  createReviewSubject,
  createDiffWindow,
  extractIssueCount,
  collectBlockingIssues,
  buildPassGateResult,
  buildFailedGateResult,
  resolveQualityReviewProfile,
  resolveCliStatePath,
  resolveExistingCliStatePath,
  getGitHead,
}

if (require.main === module) main()
